### PR TITLE
post gather module for saltstack salt

### PR DESCRIPTION
**Pull request by h00die**
This PR adds a post module to gather salt information, configs, etc.  I'm not a salt admin, so if anyone knows some other things that would be worth pulling, I'll be happy to add it.  Putting this in draft as I work with a new dev while they get the Windows portion working.

Salt is VERY dynamic, so I tried to cover a bunch of cases and pull as much data as possible and put it in the right places.  Undoubtedly some things wont work right, edge cases etc.  I think this should be considered a 'round 1' and it'll get updated as people find oddities.  Also, in the future new modules will get released which either make some calls not work, or make new calls possible.

Some notes:

- I tested on ubuntu 20.04 as the master
- ubuntu 20.04 minion
- windows server 2012 minion
- macos minion

## Verification

- [x] Start `msfconsole`
- [x] Get a shell on a box that is either a salt master or salt minion, linux or windows
- [x] `use post/multi/gather/saltstack.rb`
- [x] **Verify** you get lots of juicy info
- [x] **Document** looks pretty fly for a pentest guy

### CODE REVIEW

**Comment by wvu-r7** [comment]
Cool!

**Review by wvu — CHANGES_REQUESTED** [review]
I'd rename this `post/multi/gather/saltstack_salt`, since SaltStack is the vendor, and Salt is the product.

**Comment by h00die** [comment]
> I'd rename this `post/multi/gather/saltstack_salt`, since SaltStack is the vendor, and Salt is the product.

Don't be salty

File: modules/post/multi/gather/saltstack_salt.rb [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+require 'yaml'
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::File
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'SaltStack Information Gatherer',
```
Comment (wvu-r7) [inline modules/post/multi/gather/saltstack_salt.rb:15]:
```suggestion
        'Name' => 'SaltStack Salt Information Gatherer',
```

File: modules/post/multi/gather/saltstack_salt.rb [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        'Description' => %q{
+          This module gathers information from SaltStack masters and minions.
```
Comment (wvu-r7) [inline modules/post/multi/gather/saltstack_salt.rb:17]:
```suggestion
          This module gathers information from SaltStack Salt masters and minions.
```

**Review by wvu — CHANGES_REQUESTED** [review]
Still a few places to be consistent in naming. See `store_loot` calls, too.

**Comment by c2Vlcgo** [comment]
Runs without issue in my dev environment.  Output files match result expectations; granular breakdown of SLS trees, and a good overview of all available minion hosts.

Salt master: Ubuntu 16.04
Salt minion: Ubuntu 16.04
Salt Versions Tested: 3001 and 3002.6

Screenshow below shows execution output; IP addresses redacted.

saltstack_salt_testing

**Review by wvu — APPROVED** [review]

File: modules/post/multi/gather/saltstack_salt.rb [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    ['/etc/salt/roster'].each do |config|
+      next unless file?(config)
+
+      begin
+        minions = YAML.safe_load(read_file(config))
+      rescue Psych::SyntaxError
+        print_error("Unable to load #{config}")
+        next
+      end
+      minions.each do |name, minion|
```
Comment (bwatters-r7) [inline modules/post/multi/gather/saltstack_salt.rb:222]: You may want to guard this with 
`unless minions.nil?`
I found that the value `minions` came back as `nil` if salt-ssh was not installed on the minion:
```
[+] 192.168.134.113:60682 - minion data gathering successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513144840_default_192.168.134.113_saltstack_minion_361736.bin
[+] Found minion: ubuntu-20041 (192.168.134.113) - Linux version 5.8.0-53-generic (buildd@lcy01-amd64-012) (gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #60~20.04.1-Ubuntu SMP Thu May 6 09:52:46 UTC 2021
[*] Showing SLS
[+] 192.168.134.113:60682 - SLS output successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513144857_default_192.168.134.113_saltstack_sls_626023.txt
[*] Loading roster
[-] Post failed: NoMethodError undefined method `each' for nil:NilClass
[-] Call stack:
[-]   /home/tmoose/rapid7/metasploit-framework/modules/post/multi/gather/saltstack_salt.rb:223:in `block in master'
[-]   /home/tmoose/rapid7/metasploit-framework/modules/post/multi/gather/saltstack_salt.rb:214:in `each'
[-]   /home/tmoose/rapid7/metasploit-framework/modules/post/multi/gather/saltstack_salt.rb:214:in `master'
[-]   /home/tmoose/rapid7/metasploit-framework/modules/post/multi/gather/saltstack_salt.rb:299:in `run'
[*] Post module execution completed
```

**Comment by bwatters-r7** [comment]
Looks great.  Just the one thought about catching `nil` if salt-ssh is not installed on the minions.
```
msf6 post(multi/gather/saltstack_salt) > run

[!] SESSION may not be compatible with this module.
[*] Looking for salt minion config files
[+] Minion master: 192.168.134.113
[+] 192.168.134.113:60682 - minion file successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513165358_default_192.168.134.113_saltstack_minion_137779.bin
[*] Attempting to list minions
[*] minions:
- ubuntu-20041
- ubuntu-vm
minions_denied: []
minions_pre: []
minions_rejected: []
[+] 192.168.134.113:60682 - minion file successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513165359_default_192.168.134.113_saltstack_minion_836560.bin
[+] Minions List
============

 Status    Minion Name
 ------    -----------
 Accepted  ubuntu-20041
 Accepted  ubuntu-vm

[*] Gathering data from minions (this can take some time)
[*] ubuntu-20041:
  network.get_hostname: ubuntu-20041
  network.interfaces:
    ens32:
      hwaddr: 00:0c:29:f9:61:ef
      inet:
      - address: 192.168.134.113
        broadcast: 192.168.134.255
        label: ens32
        netmask: 255.255.255.0
      inet6:
      - address: fd34:fe56:7891:2f3b:2c25:cced:3232:fcea
        prefixlen: '64'
        scope: global
      - address: fd34:fe56:7891:2f3b:ff34:7a5f:f785:29e5
        prefixlen: '64'
        scope: global
      - address: fe80::3c97:c033:acf:f037
        prefixlen: '64'
        scope: link
      up: true
    lo:
      hwaddr: 00:00:00:00:00:00
      inet:
      - address: 127.0.0.1
        broadcast: null
        label: lo
        netmask: 255.0.0.0
      inet6:
      - address: ::1
        prefixlen: '128'
        scope: host
      up: true
  status.version: 'Linux version 5.8.0-53-generic (buildd@lcy01-amd64-012) (gcc (Ubuntu
    9.3.0-17ubuntu1~20.04) 9.3.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #60~20.04.1-Ubuntu
    SMP Thu May 6 09:52:46 UTC 2021'
  system.get_system_info: "Traceback (most recent call last):\n  File \"/usr/lib/python3/dist-packages/salt/minion.py\",
    line 2098, in _thread_multi_return\n    return_data = minion_instance._execute_job_function(\n
    \ File \"/usr/lib/python3/dist-packages/salt/minion.py\", line 1861, in _execute_job_function\n
    \   return_data = self.executors[fname](opts, data, func, args, kwargs)\n  File
    \"/usr/lib/python3/dist-packages/salt/loader.py\", line 1235, in __call__\n    return
    self.loader.run(run_func, *args, **kwargs)\n  File \"/usr/lib/python3/dist-packages/salt/loader.py\",
    line 2268, in run\n    return self._last_context.run(self._run_as, _func_or_method,
    *args, **kwargs)\n  File \"/usr/lib/python3/dist-packages/salt/loader.py\", line
    2283, in _run_as\n    return _func_or_method(*args, **kwargs)\n  File \"/usr/lib/python3/dist-packages/salt/executors/direct_call.py\",
    line 12, in execute\n    return func(*args, **kwargs)\nTypeError: 'str' object
    is not callable\n"
ubuntu-vm:
  network.get_hostname: ubuntu-vm
  network.interfaces:
    ens32:
      hwaddr: 00:0c:29:a2:ff:7f
      inet:
      - address: 192.168.134.134
        broadcast: 192.168.134.255
        label: ens32
        netmask: 255.255.255.0
      inet6:
      - address: fd34:fe56:7891:2f3b:9d32:a7c9:ec6b:95a3
        prefixlen: '64'
        scope: global
      - address: fd34:fe56:7891:2f3b:606:98c1:b25f:68aa
        prefixlen: '64'
        scope: global
      - address: fe80::e3f3:5a95:9951:f06b
        prefixlen: '64'
        scope: link
      up: true
    lo:
      hwaddr: 00:00:00:00:00:00
      inet:
      - address: 127.0.0.1
        broadcast: null
        label: lo
        netmask: 255.0.0.0
      inet6:
      - address: ::1
        prefixlen: '128'
        scope: host
      up: true
  status.version: 'Linux version 5.8.0-53-generic (buildd@lcy01-amd64-012) (gcc (Ubuntu
    9.3.0-17ubuntu1~20.04) 9.3.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #60~20.04.1-Ubuntu
    SMP Thu May 6 09:52:46 UTC 2021'
  system.get_system_info: "Traceback (most recent call last):\n  File \"/usr/lib/python3/dist-packages/salt/minion.py\",
    line 2098, in _thread_multi_return\n    return_data = minion_instance._execute_job_function(\n
    \ File \"/usr/lib/python3/dist-packages/salt/minion.py\", line 1861, in _execute_job_function\n
    \   return_data = self.executors[fname](opts, data, func, args, kwargs)\n  File
    \"/usr/lib/python3/dist-packages/salt/loader.py\", line 1235, in __call__\n    return
    self.loader.run(run_func, *args, **kwargs)\n  File \"/usr/lib/python3/dist-packages/salt/loader.py\",
    line 2268, in run\n    return self._last_context.run(self._run_as, _func_or_method,
    *args, **kwargs)\n  File \"/usr/lib/python3/dist-packages/salt/loader.py\", line
    2283, in _run_as\n    return _func_or_method(*args, **kwargs)\n  File \"/usr/lib/python3/dist-packages/salt/executors/direct_call.py\",
    line 12, in execute\n    return func(*args, **kwargs)\nTypeError: 'str' object
    is not callable\n"
[+] 192.168.134.113:60682 - minion data gathering successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513165401_default_192.168.134.113_saltstack_minion_033068.bin
[+] Found minion: ubuntu-20041 (192.168.134.113) - Linux version 5.8.0-53-generic (buildd@lcy01-amd64-012) (gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #60~20.04.1-Ubuntu SMP Thu May 6 09:52:46 UTC 2021
[+] Found minion: ubuntu-vm (192.168.134.134) - Linux version 5.8.0-53-generic (buildd@lcy01-amd64-012) (gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #60~20.04.1-Ubuntu SMP Thu May 6 09:52:46 UTC 2021
[*] Showing SLS
[+] 192.168.134.113:60682 - SLS output successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513165402_default_192.168.134.113_saltstack_sls_171640.txt
[*] Loading roster
[*] Looking for salt minion config files
[+] Minion master: 192.168.134.113
[+] 192.168.134.113:60682 - minion file successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513165402_default_192.168.134.113_saltstack_minion_183800.bin
[+] 192.168.134.113:60682 - roster file successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513165402_default_192.168.134.113_saltstack_roster_690465.bin
[*] Gathering pillar data
[*] ubuntu-20041: {}
ubuntu-vm: {}
[+] 192.168.134.113:60682 - pillar data gathering successfully retrieved and saved to /home/tmoose/.msf4/loot/20210513165403_default_192.168.134.113_saltstack_pillar_410702.bin
[*] Post module execution completed
msf6 post(multi/gather/saltstack_salt) > 

```

**Comment by h00die** [comment]
guarded!

**Comment by bwatters-r7** [comment]
# Release Notes
This PR adds a post module to gather salt information, configs, etc.

**Comment by wvu-r7** [comment]
Thanks for handling this, @bwatters-r7!

**Pull request merged**